### ISSUE TITLE
Logging/Debugging Phoenix Apps |> What are the options? |> Can we use ElasticSearch + Kibana?

### ISSUE THREAD

**Issue by nelsonic** [dwyl/learn-phoenix-framework#64]
Logging is an _incredibly_ important activity for any web application, 🔥 
but sadly, one of the things _most_ people only think about when things "_go wrong_"... 😿 

![no logs](https://cloud.githubusercontent.com/assets/194400/23974992/4e538fa6-09d6-11e7-9766-e7427d706a70.png)

Our goal is to be able to:
+ [ ] Read the logs in a _human friendly_ format without scrolling through mountains of plaintext 
+ [ ] _Search_/_Filter_ logs
+ [ ] _Correlate_/Trace all logs for a given request together (_this isn't an issue if the app has only a handful of users/requests per minute but is **essential** when we have multiple request per second and we need to "trace" a given request through the stack to understand where/when it's failing...!_)

I was _considering_ pushing the log to ElasticSearch _asynchronously_ (_not interrupting the flow of execution_) which could then be viewed in Kibana (_in the browser_).

> We used "ELK" @ TC and it worked well for tracing logs through our stack using a "correlation ID" for a given request. see: https://www.elastic.co/videos/kibana-logstash

@jruts / @tonydaly / @katbow / @SimonLab / @shouston3 / @jackcarlisle / @des-des / @Jbarget 
have you come across a good logging tool/solution for Elixir/Phoenix ?

We are going to need logging in _all_ our apps so this discussion is relevant to _everyone_ building apps with "PETE". Also having logging/tracing as a skill will _seriously_ help you "level up". 🥇 

I just came across this while trying to debug something for "dwylbot" and noted that
2e are not _currently_ logging the Webhook payloads in [/web/controllers/webhooks_controller.ex](https://github.com/dwyl/dwylbot/blob/d28c1790f820c48b62360a450c3c2d953aa71cc5/web/controllers/webhooks_controller.ex) so this is an _immediate_ pain I have ... 
but it's also a _requirement_ for all our client/internal projects ASAP! 💸

**Comment by nelsonic** [comment]
There a few options for logging: https://elements.heroku.com/addons
![image](https://cloud.githubusercontent.com/assets/194400/23985902/63807f2c-0a19-11e7-83ab-626aa267c04b.png)

has anyone got _experience_ using any of these?

**Comment by tonydaly** [comment]
I hate Logentries + Papertrail, always found them slow and hard to use. I'm a fan of the ELK stack, Kibana is good (but needs time to set it up properly). Another option is [Splunk](https://www.splunk.com/) which is great but gets pricey.

I was going to look into trying out Logz.io for Healthlocker as it's a managed ELK stack and a Heroku Add-on, so looks easy to setup.

**Comment by nelsonic** [comment]
@tonydaly yeah, Splunk is a _great_ product but the [pricing](https://www.splunk.com/en_us/products/pricing.html) is _bonkers_! 💸 🔥 
![image](https://cloud.githubusercontent.com/assets/194400/24033963/3bcc0db0-0ae7-11e7-99f2-38dbb525b04d.png)
Fairly certain this is the ***most expensive*** price-per-GB of any SaaS Service...? 😜 

Agree there is some setup required in Kibana, if you want to delegate some of that, send it over!
http://logz.io is a good shout! please create an issue on Healthlocker and log progress. thanks! 🍦 🚀

**Comment by des-des** [comment]
@nelsonic not yet! :wink:

**Comment by nelsonic** [comment]
Found https://logz.io via https://dev.bleacherreport.com/elixir-phoenix-centralized-http-logging-aa50efe3105b
![image](https://user-images.githubusercontent.com/194400/35113595-dfd52084-fc79-11e7-8bd2-87250a63871d.png)
The "Community Edition" with "Free" 3GB/day seems pretty good ... what's the "catch"...?
